src/frontend/Semantic_check.ml

```diff
@@ -170,164 +104,67 @@ let probability_distribution_name_variants id =
 
 let lub_rt loc rt1 rt2 =
   match (rt1, rt2) with
-  | Middle.ReturnType UReal, Middle.ReturnType UInt
+  | Mir.ReturnType UReal, Mir.ReturnType UInt
    |ReturnType UInt, ReturnType UReal ->
-      Middle.ReturnType UReal
-  | rt1, rt2 when rt1 = rt2 -> rt2
-  | _ ->
-      semantic_error ~loc
-        ( "Branches of function definition need to have the same return type. \
-           Instead, found return types "
-        ^ pretty_print_returntype rt1
-        ^ " and "
-        ^ pretty_print_returntype rt2
-        ^ "." )
-
-let try_compute_ifthenelse_statement_returntype loc srt1 srt2 =
-  match (srt1, srt2) with
-  | Complete rt1, Complete rt2 -> Complete (lub_rt loc rt1 rt2)
-  | Incomplete rt1, Incomplete rt2
-   |Complete rt1, Incomplete rt2
-   |Incomplete rt1, Complete rt2 ->
-      Incomplete (lub_rt loc rt1 rt2)
-  | AnyReturnType, NoReturnType
-   |NoReturnType, AnyReturnType
-   |NoReturnType, NoReturnType ->
-      NoReturnType
-  | AnyReturnType, Incomplete rt
-   |Incomplete rt, AnyReturnType
-   |Complete rt, NoReturnType
-   |NoReturnType, Complete rt
-   |NoReturnType, Incomplete rt
-   |Incomplete rt, NoReturnType ->
-      Incomplete rt
-  | Complete rt, AnyReturnType | AnyReturnType, Complete rt -> Complete rt
-  | AnyReturnType, AnyReturnType -> AnyReturnType
-
-let try_compute_block_statement_returntype loc srt1 srt2 =
-  match (srt1, srt2) with
-  | Complete rt1, Complete rt2 | Incomplete rt1, Complete rt2 ->
-      Complete (lub_rt loc rt1 rt2)
-  | Incomplete rt1, Incomplete rt2 | Complete rt1, Incomplete rt2 ->
-      Incomplete (lub_rt loc rt1 rt2)
-  | NoReturnType, NoReturnType -> NoReturnType
-  | AnyReturnType, Incomplete rt
-   |Complete rt, NoReturnType
-   |NoReturnType, Incomplete rt
-   |Incomplete rt, NoReturnType ->
-      Incomplete rt
-  | NoReturnType, Complete rt
-   |Complete rt, AnyReturnType
-   |Incomplete rt, AnyReturnType
-   |AnyReturnType, Complete rt ->
-      Complete rt
-  | AnyReturnType, NoReturnType
-   |NoReturnType, AnyReturnType
-   |AnyReturnType, AnyReturnType ->
-      AnyReturnType
+      Validate.ok (Mir.ReturnType UReal)
+  | _, _ when rt1 = rt2 -> Validate.ok rt2
+  | _ -> Semantic_error.mismatched_return_types loc rt1 rt2 |> Validate.error
 
 let check_fresh_variable_basic id is_nullary_function =
-  (* No shadowing! *)
-  (* For some strange reason, Stan allows user declared identifiers that are
+  Validate.(
```

Comment:
Could you maybe explain what this line does? Validate is a functor, right? It's Validation.Make. So I would expect it to apply to be applied to a module which contains a single type. Here, it seems to be applied to a value of result type. Is that related to the dot after Validate? Could you help me parse this? Thanks!

Reply:
At the moment, `Validation` is written so that you construct it via `Validation.Make` passing in another module which has some type `t` which is the type of error you are going to accumulate.

In `semantic_check.ml`, I create a module `Validate` by passing `Semantic_error` into `Validation.Make` fixing the type of error at `Semantic_error.t`.

When I write `Validate.( ... )`, this `Validate` module is in scope inside the brackets. This is particularly helpful when using infix functions like `>>=`; without it I would have to qualify everything.

Reply:
Oh gotcha! So would this be equivalent to writing a 
` let open Validate` first?

Reply:
Yes - there are two ways of writing local open. As well as being more lightweight, I always think this syntax mirrors what I'm doing in my head.

Reply:
Cool, thanks!